server/client.go

```diff
@@ -943,7 +965,30 @@ func (c *client) typeString() string {
 	return "Unknown Type"
 }
 
+func (c *client) closeConnectionWithEvent(reason string) {
+
+	c.mu.Lock()
+
+	if c.nc == nil {
```

Comment:
I have an issue with this function. You could have multiple calls to this function and it will try to send the notification several times. Normally, in close(), you would use something that is set to make sure that the second call is idempotent. In close() we use c.nc but in this function, c.nc will not be set to nil until much later, and after the lock is released. So you may need to use a dedicated boolean to indicate that the connection has already been closed, such as `closed bool` ;-)